cpp/modmesh/linalg/kalman_filter.hpp

```diff
@@ -0,0 +1,528 @@
+            {
+                if (i > 0)
+                {
+                    oss << ", ";
+                }
+                oss << m_b.shape(i);
+            }
+            oss << ")";
+            throw std::invalid_argument(oss.str());
+        }
+    }
+}
+
+template <typename T>
+void KalmanFilter<T>::predict_state()
+{
+    // x <- F x  (m_x <- m_f @ m_x)
+    array_type x_col = m_x.reshape(small_vector<size_t>{m_state_size, 1});
+    x_col = m_f.matmul(x_col);
+    m_x = x_col.reshape(small_vector<size_t>{m_state_size});
+}
+
+template <typename T>
+void KalmanFilter<T>::predict_state(array_type const & u)
+{
+    // x <- F x + B u  (m_x <- m_f @ m_x + m_b @ u)
+    array_type x_col = m_x.reshape(small_vector<size_t>{m_state_size, 1});
+    array_type u_col = u.reshape(small_vector<size_t>{m_control_size, 1});
+    x_col = m_f.matmul(x_col).add(m_b.matmul(u_col));
+    m_x = x_col.reshape(small_vector<size_t>{m_state_size});
+}
+
+template <typename T>
+void KalmanFilter<T>::predict_covariance()
+{
+    // P <- F P F^H + Q  (m_p <- m_f @ m_p @ m_f^H + m_q)
+    array_type f_h = hermitian(m_f);
+    m_p = m_f.matmul(m_p).matmul(f_h).add(m_q);
+}
+
+template <typename T>
+typename KalmanFilter<T>::array_type KalmanFilter<T>::symmetrize(array_type const & p)
+{
+    // P <- 0.5(P + P^H)  (p <- 0.5(p + p^H))
+    array_type p_h = p;
+    hermitian(p_h);
+    return p.add(p_h).mul(static_cast<T>(0.5));
+}
+
+template <typename T>
+void KalmanFilter<T>::check_measurement(array_type const & z)
+{
+    if (z.ndim() != 1 || z.shape(0) != m_measurement_size)
+    {
+        Formatter oss;
+        oss << "KalmanFilter::check_measurement: The measurement SimpleArray z must be 1D of length measurement_size (" << m_measurement_size << "), but got shape (";
+        for (ssize_t i = 0; i < z.ndim(); ++i)
+        {
+            if (i > 0)
+            {
+                oss << ", ";
+            }
+            oss << z.shape(i);
+        }
+        oss << ")";
+        throw std::invalid_argument(oss.str());
+    }
+}
+
+template <typename T>
+void KalmanFilter<T>::check_control(array_type const & u)
+{
+    if (m_control_size == 0)
+    {
+        Formatter oss;
+        oss << "KalmanFilter::check_control: Control input not supported: control_size is 0";
+        throw std::invalid_argument(oss.str());
+    }
+    if (u.ndim() != 1 || u.shape(0) != m_control_size)
+    {
+        Formatter oss;
+        oss << "KalmanFilter::check_control: The control SimpleArray u must be 1D of length control_size (" << m_control_size << "), but got shape (";
+        for (ssize_t i = 0; i < u.ndim(); ++i)
+        {
+            if (i > 0)
+            {
+                oss << ", ";
+            }
+            oss << u.shape(i);
+        }
+        oss << ")";
+        throw std::invalid_argument(oss.str());
+    }
+}
+
+template <typename T>
+typename KalmanFilter<T>::array_type KalmanFilter<T>::innovation(array_type const & z)
```

Comment:
This is good.  Let us follow the style to avoid `auto`.

```cpp
template <typename T>
typename KalmanFilter<T>::array_type KalmanFilter<T>::innovation(array_type const & z)
```

It is harder to write but much easier to read (clearer to parse in mind).